### PR TITLE
LPD-105501 Render activation and reset forms in the user's language

### PR DESCRIPTION
**Pull request by wanderlast**
https://liferay.atlassian.net/browse/LPD-105501

## What Is Being Fixed

The account activation email builds its link from the `$password_setup_url$` token, and that link carried a `languageId` parameter that nothing in the portal reads. Request locale is resolved by `PortalImpl.getLocale`, which consults the `I18N_LANGUAGE_ID` request attribute, the `doAsUserLanguageId` parameter, the session locale, the authenticated user, the language cookie, `Accept-Language` when `locale.default.request` is enabled, and finally the company default. A `languageId` parameter appears nowhere in that chain; it is read in exactly one place in the portal, `UpdateLanguageAction` for `/c/portal/update_language`.

The recipient opens the link as a guest, so the user-locale step is skipped and there is no session locale or language cookie on a first visit. Resolution therefore fell through to the company default and the form was always rendered in that language, whatever the user's own language was. The parameter has never had any effect: LPD-1790 added it to the URL in 2024 and added no consumer on the receiving side.

The same URL is built a second time in `sendPassword`, for the reset email behind Forgot Password. That one carried no locale parameter at all, so it had the identical defect and is fixed here too.

## How It Is Being Fixed

Both call sites now emit `doAsUserLanguageId`, a parameter `getLocale` already honours on any request and ahead of its fall-through to the company default, so no change is needed on the receiving side. It degrades gracefully: when the user's locale is not among the instance's available locales, resolution simply continues down the chain to the current behaviour, so no link can break. It also survives the GET-to-POST conversion in `UpdatePasswordAction.resendAsPost`, which copies every request parameter into hidden inputs.

The alternative was to prefix the URL with an i18n path segment, as the reporter suggested. That was rejected on two grounds. An unavailable locale makes `I18nServlet.getI18nData` return null and the link 404s, because `locale.use.default.if.not.available` defaults to false, so the prefix needs an availability guard to be safe where the parameter does not. And nothing in the codebase builds a `/<lang>/c/...` URL today — every i18n prefix is prepended to a friendly-URL servlet mapping — so it would be a new URL shape with no existing coverage.

One trade-off worth recording for review: `doAsUserLanguageId` is named for impersonation, and `UpdateLanguageAction` only appends it alongside a `doAsUserId`. It is functionally ungated — `getLocale` honours it on any request, and `doAsUserId` is validated separately — but if it were ever gated behind `doAsUserId`, these links would silently regress.

Coverage is three integration tests in `UserLocalServiceTest`, which had no coverage of these emails at all before this change. Two assert the emailed URL carries `doAsUserLanguageId` and no longer carries the inert parameter; the third takes the URL out of the sent email, requests it, and asserts the returned page renders in the user's language rather than only that the parameter is present.

## PR Check

**pr-check: PASS** — tested on `f8cb2797fbfdce5909bcdfabbfc6b801b5359cc3`

| Validation | Result |
| --- | --- |
| Service Builder | PASS |
| Source Format | PASS |
| Service Registration | PASS |
| Transaction Usage | PASS |
| Full Portal Build | NOT VERIFIED |
| Per-Module Compile | NOT VERIFIED |
| Integration Test Compile | PASS |
| Cross-Module Compile | PASS |
| Baseline | PASS (baseline-all + seven Ant projects + 0 changed exporting modules) |
| Java Unit Tests | NOT VERIFIED |

Source Format passed for this branch but the run exits nonzero on pre-existing violations outside its diff, in `modules/apps/fragment/fragment-video-streaming`: an unfixable `JavaFeatureFlagManagerUtilCheck` on `VideoStreamingFragmentRenderer.java`, plus `CopyrightCheck` on that file, `constants/VideoStreamingWebKeys.java`, `META-INF/resources/init.jsp` and `META-INF/resources/page.jsp`. All four are byte-identical between master and HEAD and the module landed on master in June 2026, so none is attributable to this branch. The formatter rewrote them in the working tree; they were restored and not committed. The branch's own two files produced no `SourceCheck` violations. The commit-message check first failed on the word "password" in all three commit titles, per `git.commit.vulnerability.keywords`; the titles were reworded and it passes. The yarn side was skipped rather than clean, since the diff contains no js, json, jsp, jspf, scss, ts or tsx file.

Full Portal Build was trimmed at the Pass 1 estimate checkpoint, where the matched set came to roughly 25 minutes, so it established nothing about this branch. Portal core was exercised instead by `ant compile install-portal-snapshots`, which Cross-Module Compile confirmed carried the branch change into the installed snapshot jar.

Per-Module Compile verified nothing. The only changed path under `modules` is `src/testIntegration`, which the validation excludes from the deploy set because Integration Test Compile covers it and `-test` modules deploy no runtime bundle, leaving the deploy set empty. Its lockfile check ran and had nothing to check, since the diff touches no `package.json`.

Java Unit Tests verified nothing. There is no unit-test counterpart for `UserLocalServiceImpl` anywhere in the tree, and no file under `portal-impl/test/unit` or `portal-kernel/test/unit` naming `UserLocalService`. The class is covered by the integration test `UserLocalServiceTest` in `apps:user:user-test` instead, which this validation does not run.

@smottal

### CODE REVIEW

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 1]
```diff
@@ -238,6 +249,47 @@ public void testAddLDAPUserWithoutLDAPPasswordPolicy() throws Exception {
...
+		MailMessage mailMessage = MailServiceTestUtil.getLastMailMessage();
+
+		String content = _postURL(_getPasswordSetupURL(mailMessage.getBody()));
+
+		Assert.assertTrue(content, content.contains("lang=\"fr-FR\""));
+		Assert.assertTrue(
+			content,
+			content.contains(
+				LanguageUtil.get(LocaleUtil.FRANCE, "change-password")));
+	}
```
Comment (smottal) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java]: cannot this be in the same test of the other one? first we check if the mail emssage contains bla bla bla ... then we get the do the post to the url and we verify the "page content"

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 2]
```diff
@@ -238,6 +249,47 @@ public void testAddLDAPUserWithoutLDAPPasswordPolicy() throws Exception {
...
+
+	@Test
+	@TestInfo("LPD-105501")
+	public void testAddUserSendsPasswordSetupURLWithUserLanguageId()
```
Comment (smottal) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java]: testAddUserWithUserLanguageId

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 3]
```diff
@@ -1382,6 +1434,38 @@ public void testSearchUsersFromDatabase() throws Exception {
 		}
 	}
 
+	@Test
+	@TestInfo("LPD-105501")
+	public void testSendPasswordWithUserLanguageId() throws Exception {
+		_user = UserTestUtil.addUser(
+			TestPropsValues.getGroupId(), LocaleUtil.FRANCE);
```
Comment (smottal) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java:1431]: why here we can create the user like this and in the other test we need to create with a "custom" method?

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 4]
```diff
@@ -1962,6 +2046,31 @@ private long[] _addUsers(int numberOfUsers) throws Exception {
 		return userIds;
 	}
 
+	private User _addUserWithPasswordSetupEmail(Locale locale)
```
Comment (smottal) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java]: _addUser

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 5]
```diff
@@ -1962,6 +2046,31 @@ private long[] _addUsers(int numberOfUsers) throws Exception {
...
+		throws Exception {
+
+		ServiceContext serviceContext =
+			ServiceContextTestUtil.getServiceContext();
+
+		serviceContext.setPathMain(_portal.getPathMain());
+
+		String screenName = RandomTestUtil.randomString(
```
Comment (smottal) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java:2033]: inline

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 6]
```diff
@@ -1962,6 +2046,31 @@ private long[] _addUsers(int numberOfUsers) throws Exception {
...
+			NumericStringRandomizerBumper.INSTANCE,
+			UniqueStringRandomizerBumper.INSTANCE);
+
+		String emailAddress =
```
Comment (smottal) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java]: inline

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 7]
```diff
@@ -2008,6 +2117,29 @@ private void _assertUserHasPasswordPolicy(boolean ldapUser, User user)
...
+		if (!matcher.find()) {
+			throw new IllegalStateException(
+				"Unable to find a password setup URL in " + body);
+		}
+
+		return matcher.group();
+	}
+
+	private String _postURL(String url) throws Exception {
+		HttpURLConnection httpURLConnection = (HttpURLConnection)new URL(
```
Comment (smottal) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java]: no chain

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java [hunk 8]
```diff
@@ -2397,7 +2529,31 @@ private SafeCloseable _updateSecurityWithSafeCloseable(
 			company.isSiteLogo());
 	}
 
+	private SafeCloseable _updateSendPasswordResetLinkWithSafeCloseable(
```
Comment (smottal) [inline modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java]: _updateSecuritySafeCloseable ... or what it the standard name for the SafeCloseable?

File: portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java [hunk 9]
```diff
@@ -4136,8 +4136,8 @@ public boolean sendPassword(
 			passwordResetURL = StringBundler.concat(
 				serviceContext.getPortalURL(), serviceContext.getPathMain(),
 				"/portal/update_password?p_l_id=", serviceContext.getPlid(),
-				"&ticketId=", ticket.getTicketId(), "&ticketKey=",
-				ticket.getKey());
+				"&doAsUserLanguageId=", user.getLanguageId(), "&ticketId=",
```
Comment (smottal) [inline portal-impl/src/com/liferay/portal/service/impl/UserLocalServiceImpl.java]: not sure but I think this should go before p_l_id and "do < p_"

**Review by smottal — CHANGES_REQUESTED** [review]

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java (hunk 3 above)
Comment (wanderlast) [reply modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java:1431]: UserTestUtil.addUser() sets sendMail to false which means we would never get the activation email for the other test that we are actually testing against. This test doesn't rely on that email and doesn't check against sendMail so we can use the pre-existing addUser methods.

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java (hunk 5 above)
Comment (wanderlast) [reply modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java:2033]: Formatter rejects this one when I try to inline both screenName and emailAddress:

> Create a new var for "RandomTestUtil.randomString" for better readability:
  .../UserLocalServiceTest.java 2036 (SourceCheck:JavaTermStylingCheck)

File: modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java (hunk 8 above)
Comment (wanderlast) [reply modules/apps/user/user-test/src/testIntegration/java/com/liferay/user/service/test/UserLocalServiceTest.java]: Since the method above is already _updateSecurityWithSafeCloseable(long, boolean) but the boolean corresponds specifically to the "strangersverify" boolean, I've renamed both to disambiguate but to make it more similar to the updateSecurity naming pattern:

> _updateSecurityStrangersVerifyWithSafeCloseable
_updateSecuritySendPasswordResetLinkWithSafeCloseable

Hopefully that's okay, but if you prefer me keep the other and disambiguate in a different way, please let me know and I'll take a look.

**Comment by wanderlast** [comment]
**pr-check: PASS** — tested on `ae95fdc94593134b02fd4cf497dd24ce696385b5`

| Validation | Result |
| --- | --- |
| Service Builder | PASS |
| Source Format | PASS |
| Service Registration | PASS |
| Transaction Usage | PASS |
| Full Portal Build | PASS |
| Per-Module Compile | NOT VERIFIED |
| Integration Test Compile | PASS |
| Cross-Module Compile | PASS |
| Baseline | PASS (baseline-all + seven Ant projects + 1 changed exporting module) |
| Java Unit Tests | NOT VERIFIED |

Service Builder regenerated identical output — `ant build-services` left no drift, so nothing was committed. Neither changed file is a Service Builder input.

Source Format passed clean, a change from the previous run: master has moved on and the `fragment-video-streaming` violations that run had to work around are gone, so no file needed restoring. The yarn side was skipped, not clean — the diff is two `.java` files, which match no extension in `source.formatter.frontend.file.regex`, so `skip.node.task` was set and the yarn formatter never ran. Commit-message validation did run; an `ant -v` rerun confirmed `validate.commit.messages=true` and `git.working.branch.name=liferay-commerce/master` reached `SourceFormatter`.

Full Portal Build passed this time rather than being trimmed: `ant all` completed `BUILD SUCCESSFUL` in 9m44s with portal-impl recompiling 2234 sources and no `error:` line anywhere in the 888-line log.

Per-Module Compile verified nothing. The only changed path under `modules` is `src/testIntegration` in a `-test` module, excluded from the deploy set because Integration Test Compile covers it and `-test` modules deploy no runtime bundle. The `portal-impl` change triggered no consumer expansion — its diff adds and removes no `public` or `protected` line. Its lockfile check ran and had nothing to check, since the diff touches no `package.json`.

Integration Test Compile compiled `apps:user:user-test` with `--rerun`; the task line is bare rather than `UP-TO-DATE` or `FROM-CACHE`, and javac's deprecation notes name the changed test file by absolute path, so the branch source was the actual compile input.

Cross-Module Compile found `apps:user:user-test` on the testIntegration surface and nothing on the deprecated surface. The same executed compile covers both searched type names, since `UserLocalServiceTest.java` imports `UserLocalServiceImpl` directly.

Java Unit Tests verified nothing. No `UserLocalServiceImplTest.java` exists anywhere in the tree, and `portal-impl/test/unit/com/liferay/portal/service/impl/` holds only three unrelated tests. The class is covered by the integration test `UserLocalServiceTest` in `apps:user:user-test`, which this branch extends and which this validation does not run.

**Review by smottal — APPROVED** [review]

**Comment by smottal** [comment]
ci:forward
